File: legacy/java/tests/StrataPersistenceTest/src/strata/persistence/testdomain/OrganizationRepositoryTest.java
package strata.persistence.testdomain;


import org.junit.After;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Ignore;
import org.junit.Test;
import strata.persistence.namedquery.INamedQuery;
import strata.persistence.unitofwork.IUnitOfWork;
import strata.persistence.unitofwork.IUnitOfWorkProvider;

public abstract 
class OrganizationRepositoryTest
{
    private IUnitOfWorkProvider      itsProvider;
    private IOrganizationRepository  itsTarget;
    
    @Before
    public void 
    setUp() throws Exception
    {
        itsProvider = createUnitOfWorkProvider();
        itsTarget = new OrganizationRepository(itsProvider);
    }

    @After
    public void 
    tearDown() throws Exception
    {
        INamedQuery<IOrganization> query = itsTarget.getNamedQuery( "GetAll" );
        IUnitOfWork      unitOfWork = itsProvider.getUnitOfWork();
        
        for (IOrganization p : query.getAll())
            itsTarget.removeOrganization( p );
        
        unitOfWork.commit();
        itsTarget = null;
        itsProvider = null;
    }

    @Test
    public void 
    testInsertOrganization() 
        throws Exception
    {
        IUnitOfWork   unitOfWork = itsProvider.getUnitOfWork();
        IOrganization expected = new Organization("FooBar Inc.");
        IOrganization actual = null;
        
        expected = itsTarget.insertOrganization( expected );
        unitOfWork.commit();
        
        actual   = itsTarget.getOrganizationByPartyKey( expected.getPartyKey() );
        Assert.assertEquals( expected,actual );
    }

    @Test
    public void 
    testUpdateOrganization() 
        throws Exception
    {
        IUnitOfWork   unitOfWork = itsProvider.getUnitOfWork();
        IOrganization expected = new Organization("FooBar Inc.");
        IOrganization actual = null;
        
        expected = itsTarget.insertOrganization( expected );
        unitOfWork.commit();
        
        unitOfWork = itsProvider.getUnitOfWork();
        expected.getMemberIds().add( 10L );
        actual = itsTarget.updateOrganization( expected );
        unitOfWork.commit();
        Assert.assertEquals( expected,actual );
    }

    @Test
    public void 
    testRemoveOrganization() 
        throws Exception
    {
        IUnitOfWork   unitOfWork = itsProvider.getUnitOfWork();
        IOrganization expected = new Organization("FooBar Inc.");

        expected = itsTarget.insertOrganization( expected );
        unitOfWork.commit();
        
        unitOfWork = itsProvider.getUnitOfWork();
        Assert.assertTrue( itsTarget.hasOrganizationWithPartyKey( expected.getPartyKey() ) );
        itsTarget.removeOrganization( expected );
        unitOfWork.commit();
        
        Assert.assertFalse( itsTarget.hasOrganizationWithPartyKey( expected.getPartyKey() ) );
    }

    @Test
    public void 
    testGetOrganizationByPartyKey()
        throws Exception
    {
        IUnitOfWork   unitOfWork = itsProvider.getUnitOfWork();
        IOrganization expected = new Organization("FooBar Inc.");
        IOrganization actual = null;
        
        expected = itsTarget.insertOrganization( expected );
        unitOfWork.commit();
        
        Assert.assertTrue( itsTarget.hasOrganizationWithPartyKey( expected.getPartyKey() ) );
        actual = itsTarget.getOrganizationByPartyKey( expected.getPartyKey() );
        Assert.assertEquals( expected,actual );
    }

    @Ignore
    @Test
    public void 
    testGetOrganizationByName() 
        throws Exception
    {
        IUnitOfWork   unitOfWork = itsProvider.getUnitOfWork();
        IOrganization expected = new Organization("FooBar Inc.");
        IOrganization actual = null;
        
        expected = itsTarget.insertOrganization( expected );
        unitOfWork.commit();
        
        Assert.assertTrue( itsTarget.hasOrganizationWithPartyKey( expected.getPartyKey() ) );
        actual = itsTarget.getOrganizationByName( expected.getName() );
        Assert.assertEquals( expected,actual );
        
    }

    @Test
    public void 
    testGetFinder()
    {
        INamedQuery<IOrganization> finder = itsTarget.getNamedQuery( "GetByName" );
        
        Assert.assertNotNull( finder );
        Assert.assertEquals( "strata.persistence.testdomain.IOrganization.GetByName",finder.getName() );
    }

    @Test
    public void 
    testHasOrganizationWithPartyKey() 
        throws Exception
    {
        IUnitOfWork   unitOfWork = itsProvider.getUnitOfWork();
        IOrganization expected = new Organization("FooBar Inc.");
        
        expected = itsTarget.insertOrganization( expected );
        unitOfWork.commit();
        
        Assert.assertTrue( itsTarget.hasOrganizationWithPartyKey( expected.getPartyKey() ) );
    }

    @Test
    public void 
    testHasOrganizationWithName() 
        throws Exception
    {
        IUnitOfWork   unitOfWork = itsProvider.getUnitOfWork();
        IOrganization expected = new Organization("FooBar Inc.");
        
        expected = itsTarget.insertOrganization( expected );
        unitOfWork.commit();
        
        Assert.assertTrue( itsTarget.hasOrganizationWithName( expected.getName() ) );
    }

    @Test
    public void 
    testHasFinder()
    {
        Assert.assertTrue( itsTarget.hasNamedQuery( "GetByName" ) );
    }

    protected abstract IUnitOfWorkProvider
    createUnitOfWorkProvider();
}

// ##########################################################################

File: csharp/Tests/Strata.Domain.CoreTest/Repository/OrganizationRepositoryTest.cs
//  ##########################################################################
//  # File Name: OrganizationRepositoryTest.cs
//  ##########################################################################

using Strata.Domain.Core.TestDomain;
using Strata.Domain.Core.UnitOfWork;
using System;
using System.Threading.Tasks;
using NUnit.Framework;
using NUnit.Framework.Internal;
using Strata.Domain.Core.NamedQuery;

namespace Strata.Domain.Core.Repository
{
    //////////////////////////////////////////////////////////////////////////
    /// <summary>
    /// </summary>
    ///  
    [TestFixture]
    public abstract
    class OrganizationRepositoryTest
    {
        private IUnitOfWorkProvider itsProvider;
        private IOrganizationRepository itsTarget;

        [SetUp]
        public void
        SetUp()
        {
            itsProvider = CreateUnitOfWorkProvider();
            itsTarget = new OrganizationRepository(itsProvider);
        }

        [TearDown]
        public async Task
        TearDown()
        {
            INamedQuery<IOrganization> query = await (itsTarget.GetNamedQuery("GetAll"));
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();

            foreach (IOrganization p in await (query.GetAll()))
                await (itsTarget.Remove(p));

            await (unitOfWork.Commit());
            itsTarget = null;
            itsProvider = null;
        }

        [Test]
        public async Task
        TestInsertOrganization()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IOrganization expected = new Organization("FooBar Inc.");
            IOrganization actual = null;

            expected = await (itsTarget.Insert(expected));
            await (unitOfWork.Commit());

            actual = (await itsTarget.GetUnique(expected.PrimaryId)).Get();
            Assert.AreEqual(expected,actual);
        }

        [Test]
        public async Task
        TestUpdateOrganization()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IOrganization expected = new Organization("FooBar Inc.");
            IOrganization actual = null;

            expected = await (itsTarget.Insert(expected));
            await (unitOfWork.Commit());

            unitOfWork = itsProvider.GetUnitOfWork();
            expected.MemberIds.Add(10L);
            actual = await (itsTarget.Update(expected));
            await (unitOfWork.Commit());
            Assert.AreEqual(expected,actual);
        }

        [Test]
        public async Task
        TestRemoveOrganization()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IOrganization expected = new Organization("FooBar Inc.");

            expected = await (itsTarget.Insert(expected));
            await (unitOfWork.Commit());

            unitOfWork = itsProvider.GetUnitOfWork();
            Assert.IsTrue(await itsTarget.HasUnique(expected.PrimaryId));
            await (itsTarget.Remove(expected));
            await (unitOfWork.Commit());

            Assert.IsFalse(await itsTarget.HasUnique(expected.PrimaryId));
        }

        [Test]
        public async Task
        TestGetOrganization()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IOrganization expected = new Organization("FooBar Inc.");
            IOrganization actual = null;

            expected = await (itsTarget.Insert(expected));
            await (unitOfWork.Commit());

            Assert.IsTrue(await itsTarget.HasUnique(expected.PrimaryId));
            actual = (await itsTarget.GetUnique(expected.PrimaryId)).Get();
            Assert.AreEqual(expected,actual);
        }

        [Ignore("")]
        [Test]
        public async Task
        TestGetOrganizationByName()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IOrganization expected = new Organization("FooBar Inc.");
            IOrganization actual = null;

            expected = await (itsTarget.Insert(expected));
            await (unitOfWork.Commit());

            Assert.IsTrue(await (itsTarget.HasUnique(expected.PrimaryId)));
            actual = (await itsTarget.GetUniqueMatching(org => org.Name.Equals(expected.Name))).Get();
            Assert.AreEqual(expected,actual);

        }

        [Test]
        public async Task
        TestGetNamedQuery()
        {
            INamedQuery<IOrganization> finder = await itsTarget.GetNamedQuery("GetByName");

            Assert.NotNull(finder);
            Assert.AreEqual("GetByName",finder.Name);
        }

        [Test]
        public async Task
        TestHasOrganizationWithPartyKey()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IOrganization expected = new Organization("FooBar Inc.");

            expected = await (itsTarget.Insert(expected));
            await (unitOfWork.Commit());

            Assert.IsTrue(await (itsTarget.HasUnique(expected.PrimaryId)));
        }

        [Test]
        public async Task
         TestHasOrganizationWithName()
        {
            IUnitOfWork unitOfWork = itsProvider.GetUnitOfWork();
            IOrganization expected = new Organization("FooBar Inc.");

            expected = await (itsTarget.Insert(expected));
            await (unitOfWork.Commit());

            Assert.IsTrue(await (itsTarget.HasMatching(org => org.Name.Equals(expected.Name))));
        }

        [Test]
        public async Task
        testHasFinder()
        {
            Assert.IsTrue(await (itsTarget.HasNamedQuery("GetByName")));
        }

        protected abstract IUnitOfWorkProvider
        CreateUnitOfWorkProvider();
    }
}

//  ##########################################################################
